File: server/internal/ports/backtest.go
package ports

import (
	"context"

	"github.com/EnsoTech-Studio/CryptoBot/server/internal/domain/backtest"
	"github.com/EnsoTech-Studio/CryptoBot/server/internal/domain/market"
)

type BacktestEngine interface {
	Run(context.Context, backtest.ExperimentSnapshot, []market.Candle, []market.BBO) (backtest.Result, error)
}

File: py_backend/app/ports/backtest.py
"""Backtest engine seam. Mirrors `server/internal/ports/backtest.go`."""

from __future__ import annotations

from typing import Protocol

from ..domain.backtest import ExperimentSnapshot, Result
from ..domain.market import BBO, Candle


class BacktestEngine(Protocol):
    def run(self, snapshot: ExperimentSnapshot, candles: list[Candle], bbo: list[BBO]) -> Result: ...
